controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+// 3. Set VRG in the preferredCluster to secondary
+// 4. Restore PV to failoverCluster
+// 5. Update UserPlacementRule decision to failoverCluster
+// 6. Create VRG for the failoverCluster as Primary
+// 7. Update AVR status
+// 8. Delete VRG MW from preferredCluster once the VRG state has changed to Secondary
+//
+func (a *AVRInstance) runFailover() (bool, error) {
+	a.log.Info("Entering runFailover", "state", a.instance.Status.LastKnownDRState)
+
+	const done = true
+
+	// We are done if empty
+	if a.instance.Spec.FailoverCluster == "" {
+		return done, fmt.Errorf("failover cluster not set. FailoverCluster is a mandatory field")
 	}
 
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		// Skip this subscription if a manifestwork already exist for it
-		mwName := BuildManifestWorkName(subscription.Name, subscription.Namespace, MWTypeVRG)
+	newHomeCluster := a.instance.Spec.FailoverCluster
 
-		mw, err := a.findManifestWork(mwName, d.HomeCluster)
-		if err != nil {
-			a.log.Error(err, "findManifestWork()", "name", subscription.Name)
+	// We are done if we have already failed over
+	if len(a.userPlacementRule.Status.Decisions) > 0 {
+		if a.instance.Spec.FailoverCluster == a.userPlacementRule.Status.Decisions[0].ClusterName {
+			a.log.Info("Already failed over", "state", a.instance.Status.LastKnownDRState)
+
+			err := a.ensureCleanup()
+			if err != nil {
+				return !done, err
+			}
 
-			return false, err
+			return done, nil
 		}
+	}
 
-		if mw != nil {
-			a.log.Info(fmt.Sprintf("Mainifestwork exists for subscription %s (%v)", subscription.Name, mw))
+	// Make sure we record the state that we are failing over
+	a.setDRState(rmn.FailingOver)
 
-			return true, nil
-		}
+	// Save the current home cluster
+	curHomeCluster := ""
+	if len(a.userPlacementRule.Status.Decisions) > 0 {
+		curHomeCluster = a.userPlacementRule.Status.Decisions[0].ClusterName
 	}
 
-	return false, nil
+	// Set VRG in the failed cluster (preferred cluster) to secondary
+	err := a.updateVRGStateToSecondary(curHomeCluster)
+	if err != nil {
+		a.log.Error(err, "Failed to update existing VRG manifestwork to secondary")
+
+		return !done, err
+	}
+
+	result, err := a.runPlacementTask(newHomeCluster, "", curHomeCluster)
+	if err != nil {
+		return !done, err
+	}
+
+	a.advanceToNextDRState()
+	a.log.Info("Exiting runFailover", "state", a.instance.Status.LastKnownDRState)
+
+	return result, nil
 }
 
-func (a *AVRInstance) findManifestWork(
-	mwName, homeCluster string) (*ocmworkv1.ManifestWork, error) {
-	if homeCluster != "" {
-		mw := &ocmworkv1.ManifestWork{}
+//
+// runFailback:
+// 1. If preferredCluster not set, get it from AVR status
+// 2. If still empty, fail it
+// 3. If the preferredCluster is the failoverCluster, fail it
+// 4. If preferredCluster is the same as the userPlacementRule decision, do nothing
+// 5. If the VRG replication state has not transitioned to secondary yet, then wait
+// 6. Set VRG in the peer cluster to secondary
```

Comment:
I see that setting the userPlacement status to no clusters is an unnecessary step.  This is the confusing part.  The way I look at it is like this:
1. update VRG state to secondary
2. VRG is in the process of changing to secondary, but in order to complete that step, some prerequisites has to be complete (the prerequisites can be PVCs get deleted)
3. Once the prerequisites have been complete, then change to secondary and report that in the VRG status.

Reply:
Combining this [comment](https://github.com/RamenDR/ramen/pull/69#discussion_r651028459) as well here as they sort of go together.

1. Deleting VRG:
    - VRG would be deleted based on the state set in VRG (primary or secondary)
    - Primary deletion will ensure that the PVs are also garbage collected, IOW we delete VRG as primary when the application is deleted form the hub (app no longer required)
    - Secondary deletion will ensure status of PVCs->VRs is set to Secondary in a safe manner (not in use)

As a result deleting VRG will not ensure moving to Secondary and then deleting VRG, it will behave as per the `vrg.spec.state` that is set. To ensure VRG state is set to "Secondary" prior to deletion, we would need to set it in the VRG-MW, ensure it reached the MC, and then delete it. For example we set to Secondary and it does not reach the MC, but delete reaches it first, then we would delete it as Primary and garbage collect the backend resources.

So for the above to be ensured we need to move VRG to Secondary, ensure MW is applied, then delete the MW (the last step will ensure VRG prior to deletion marks and ensures all VRs are Secondary). The caveat maybe that if we delete the VRG-MW we may not be able to check if VRG reached Secondary prior to rolling out the Subscriptions to the failback cluster, so we should wait for VRG to reach Secondary and then take the action of updating the userPl (more below).

2. Removing Subscriptions prior to marking VRG Secondary (or, VRG holding back on marking VRs as Secondary till PVs are in use)
    - PVCs can be in use because pods have it mounted, so VR to Secondary ensures PV is not in use
    - PVC can potentially be used subsequent to or in parallel to it currently not in use, if a pod is being rescheduled or starting up (race condition between us checking PVC is not in use and it becoming subsequently used by a pod)
    - For the 2 reasons above, VRG will ensure a PVC is not in use and also deleted (PVCs that are deleted may still be in use and will be garbage collected post they are not in use any longer)
    - For the conditions to be met above, we need the Subscriptions deleted, which in turn deletes the PVC and also the pods using the PVC from the MC, thus ensuring the volume is not in use and is dormant and can be marked Secondary for the final sync

Now the additional constraint is:
- VR will attempt to move a volume to `Primary` gracefully if possible and if not force it to `Primary` automatically
- To ensure the force action is not taken, we need to ensure remote end is marked `Secondary` and is actually `Secondary` before marking VRG (and hence VR) on the failback site as `Primary`
- Also, VR updating the volume as Primary or Secondary and PVC being mounted by pods (once Subscriptions hit the failback cluster) happen in parallel, which may (or may not) cause pod startup errors (if former then at some point it will recover, if latter it may just mount it as read-only)

As a result overall,
- Remove Subscriptions (or update userPl to no cluster)
- Mark VRG Secondary
    - Above steps can be interchanged as well (VRG will not mark PVCs Secondary anyway till Subscriptions are removed)
- Wait for VRG to reach Secondary
- Delete VRG MW (can be asynchronous, more a cleanup action)
- Change userPl to failback cluster
- Create/Update VRG on failback cluster as Primary (VRG needs PVCs to connect VRs to the same and mark as Primary, hence Subscription has to be rolled out for VRG to ensure all PVCs are present)
    - This last step needs some testing to ensure pods do not start using PVCs even when they are Secondary (which if it does will need a fix in ceph-csi to deny mount if volume is Secondary or some such)